### PR TITLE
Added list workflows API and Pagination Support

### PR DESCRIPTION
**Pull request by calum-stripe**
# Summary
Adds the `list_workflow_executions` API to the client (which support searching by the query string [List Filters API](https://docs.temporal.io/docs/temporal-explained/visibility/#list-filters)) and exposes the pagination for `list_open_workflow_executions` and `list_closed_workflow_exceptions`.

```ruby
# fully backwards compatible (executions is a custom class, but has Enumerable interface, so fully compatible with the current version)
executions = client.list_workflow_executions(namespace, query)
executions.each { |execution| ... } # iterates through all executions on all pages, loaded lazily

# limiting results per page
executions = client.list_workflow_executions(namespace, query, max_page_size: 100)
executions.each { |execution| ... } # iterates through a single page only
token = executions.next_page_token # gives you access to the token

# with per page limit and a next page token
executions = client.list_workflow_executions(namespace, query, max_page_size: 100, next_page_token: token)
executions.each { |execution| ... } # iterates through a single page only
executions.next_page_token # gives you access to the next token

# even supports next page iterator
executions = executions.next_page # gives you a next page iterator
```

# Test Plan
```
# unit test
bundle exec rspec spec/

# integration test
cd examples
(terminal 1) ./bin/worker
(terminal 2) USE_ENCRYPTION=1 ./bin/worker
(terminal 3) bundle exec rspec spec/integration
```

### CODE REVIEW

**Review by antstorm — COMMENTED** [review]
Thank you, @calum-stripe !

Can you please explain how would someone use these methods with a block? As far as I understand there's no way of just getting a single page back. And in the case of passing a block — what would you do with the `next_page_token`?

Back when we talked about it I thought that you might wanna return back a lazy iterator itself. Similar to `iterator = [1,2,3].each`. Then you can manually step through the results and fetch the next page's iterator if you need it

**Comment by calum-stripe** [comment]
Hey @antstorm. 

First of all, sorry we had some miscommunication around this. I'm also pretty new to Ruby (< 4 months) so learning lots here too.

Yeah the ergonomics here a little odd because we want to support just returning everything and no breaking API changes. The way I've been testing and using the pagination here is the following:

1. The client (in this case, our frontend web app) initially makes a request to the backend API service with no `next_page_token`. 
2. The backend service calls the `temporal-ruby` code like this (not exactly this code but more or less similar):
```ruby
...
client.list_workflow_executions(namespace, query, next_page_token: nil) do |executions, next_page_token_result| 
    result = {}
    result[:executions] = executions
    result[:next_page_token] = next_page_token_result
    return result
end
```
This returns back the first page in the list because we return in the first block
3. When the frontend app is ready to receive more, it takes the `next_page_token` and sends the same `query` and `next_page_token` to the backend API service again which now calls the `temporal-ruby` function like this:
```ruby
...
client.list_workflow_executions(namespace, query, next_page_token: next_page_token) do |executions, next_page_token_result| 
    result = {}
    result[:executions] = executions
    result[:next_page_token] = next_page_token_result
    return result
end
```
This returns back the next page in the sequence! 
4. Rinse and repeat step 3 as many times as needed (or until next_page_token is nil from the backend service)

So the long story short is, the underlying function in `temporal-ruby` is doing the heavy lifting if you allow the callback to be called multiple times (it's keeping the `next_page_token` in memory inside the lower function), we're just using it in a weird way where we store the `next_page_token` in the frontend so we can do pagination easier! 

I guess if you wanted a different number of results or more pages, you could add a counter and track this (inside the loop then return when you're done) or you can use the new `max_page_size` option i added!

**Comment by antstorm** [comment]
@calum-stripe sorry for a long response. The block approach doesn't really add much value since you're just using it to pass the executions and the next page token back. I think what would work here is something quite simple as:

```ruby
# fully backwards compatible (executions is a custom class, but has Enumerable interface, so fully compatible with the current version)
executions = client.list_workflow_executions(namespace, query)
executions.each { |execution| ... } # iterates through all executions on all pages, loaded lazily

# limiting results per page
executions = client.list_workflow_executions(namespace, query, max_page_size: 100)
executions.each { |execution| ... } # iterates through a single page only
token = executions.next_page_token # gives you access to the token

# with per page limit and a next page token
executions = client.list_workflow_executions(namespace, query, max_page_size: 100, next_page_token: token)
executions.each { |execution| ... } # iterates through a single page only
executions.next_page_token # gives you access to the next token

# as an extra bonus
executions = executions.next_page # gives you a next page iterator
```

Does this look sane to you? Please let me know if it doesn't work for any reason

**Comment by calum-stripe** [comment]
@antstorm thanks for the feedback, i implemented what you mentioned above and admit this is much better. What do you think?

**Comment by calum-stripe** [comment]
@antstorm thanks for the feedback, i implemented what you mentioned above and admit this is much better. What do you think?

File: spec/unit/lib/temporal/client_spec.rb [hunk 1]
```diff
@@ -832,42 +831,107 @@ class NamespacedWorkflow < Temporal::Workflow
...
+      end
     end
 
     context 'when given unsupported filter' do
       let(:filter) { { foo: :bar } }
 
       it 'raises ArgumentError' do
         expect do
-          subject.list_open_workflow_executions(namespace, from, filter: filter)
+          subject.list_open_workflow_executions(namespace, from, filter: filter).each
```
Comment (antstorm) [inline spec/unit/lib/temporal/client_spec.rb:934]: nit: even though calling `#each` works here, it's more idiomatic to call `#to_a` (`Enumerable` should handle it for us). The thing is that at some point we might wanna return the actual Enumerator from `#each` (this is extra for this PR, so no need to implement it now)

**Review by antstorm — CHANGES_REQUESTED** [review]

File: lib/temporal/connection/grpc.rb [hunk 2]
```diff
@@ -411,10 +411,10 @@ def terminate_workflow_execution(
         client.terminate_workflow_execution(request)
       end
 
-      def list_open_workflow_executions(namespace:, from:, to:, next_page_token: nil, workflow_id: nil, workflow: nil)
+      def list_open_workflow_executions(namespace:, from:, to:, next_page_token: nil, workflow_id: nil, workflow: nil, max_page_size: nil)
         request = Temporal::Api::WorkflowService::V1::ListOpenWorkflowExecutionsRequest.new(
           namespace: namespace,
-          maximum_page_size: options[:max_page_size],
+          maximum_page_size: max_page_size.nil? ? options[:max_page_size] : max_page_size,
```
[the remark this answers is not in the document]
Comment (calum-stripe) [reply lib/temporal/connection/grpc.rb:417]: it's set to 100 i believe, what do you mean frustrating? theres no change in behavior unless the user defines a different `max_page_size` (which then i guess they're responsible for all the extra network calls etc)

File: lib/temporal/client.rb [hunk 3]
```diff
@@ -386,16 +387,20 @@ def get_workflow_history(namespace:, workflow_id:, run_id:)
...
 
-    def list_closed_workflow_executions(namespace, from, to = Time.now, filter: {})
+    def list_closed_workflow_executions(namespace, from, to = Time.now, filter: {}, next_page_token: nil, max_page_size: nil)
       validate_filter(filter, :status, :workflow, :workflow_id)
 
-      fetch_executions(:closed, { namespace: namespace, from: from, to: to }.merge(filter))
+      Temporal::Workflow::Executions.new(connection: connection, status: :closed, request_options: { namespace: namespace, from: from, to: to, next_page_token: next_page_token, max_page_size: max_page_size}.merge(filter))
+    end
+
+    def list_workflow_executions(namespace, query, next_page_token: nil, max_page_size: nil)
```
[the remark this answers is not in the document]
Comment (calum-stripe) [reply lib/temporal/client.rb:402]: not against this so changed it on the client, do you think worth changing on the grpc too?

**Review by antstorm — APPROVED** [review]
Looks good!

**Pull request merged**